src/aiida_quantumespresso/parsers/base.py

```diff
@@ -3,15 +3,78 @@
 
 All `Parser` implementations in `aiida-quantumespresso` must use this base class, not `aiida.parsers.Parser`.
 """
-from aiida.parsers import Parser as _BaseParser
+import abc
+import re
+from typing import Optional, Tuple
 
-__all__ = ('Parser',)
+from aiida.common import AttributeDict
+from aiida.engine import ExitCode
+from aiida.parsers import Parser
 
+from aiida_quantumespresso.parsers.parse_raw.base import convert_qe_time_to_sec
 
-class Parser(_BaseParser):  # pylint: disable=abstract-method
-    """Custom `Parser` class for `aiida-quantumespresso` parser implementations."""
+__all__ = ('BaseParser',)
 
-    def emit_logs(self, logging_dictionaries, ignore=None):
+
+class BaseParser(Parser, metaclass=abc.ABCMeta):
+    """Custom ``Parser`` class for ``aiida-quantumespresso`` parser implementations."""
+
+    class_error_map = {}
+    class_warning_map = {}
+
+    base_error_map = {
+        'Maximum CPU time exceeded': 'ERROR_OUT_OF_WALLTIME',
+    }
+    base_warning_map = {
+        'Warning:': None,
+        'DEPRECATED:': None,
+    }
+    success_string = 'JOB DONE'
+
+    @classmethod
+    def get_error_map(cls):
+        """The full error map of the parser class."""
+        error_map = cls.base_error_map.copy()
+        error_map.update(cls.class_error_map)
+        return error_map
+
+    @classmethod
+    def get_warning_map(cls):
+        """The full warning map of the parser class."""
+        warning_map = cls.base_warning_map.copy()
+        warning_map.update(cls.class_warning_map)
+        return warning_map
+
+    def parse_stdout_from_retrieved(self, logs: AttributeDict) -> Tuple[str, dict, AttributeDict]:
+        """Read and parse the ``stdout`` content of a Quantum ESPRESSO calculation.
+
+        :param logs: Logging container that will be updated during parsing.
+        :returns: size 3 tuple: (``stdout`` content, parsed data, updated logs).
+        """
+        filename_stdout = self.node.get_option('output_filename')
+
+        if filename_stdout not in self.retrieved.base.repository.list_object_names():
+            logs.error.append('ERROR_OUTPUT_STDOUT_MISSING')
+            return '', {}, logs
+
+        try:
+            with self.retrieved.open(filename_stdout, 'r') as handle:
+                stdout = handle.read()
+        except OSError as exception:
+            logs.error.append('ERROR_OUTPUT_STDOUT_READ')
+            logs.error.append(exception)
+            return '', {}, logs
+
+        try:
+            parsed_data, logs = self._parse_stdout_base(stdout, logs)
+        except Exception as exception:
+            logs.error.append('ERROR_OUTPUT_STDOUT_PARSE')
+            logs.error.append(exception)
+            return stdout, {}, logs
+
+        return stdout, parsed_data, logs
+
+    def emit_logs(self, logs: AttributeDict, ignore: list = None) -> None:
```

Comment:
Incorrect type, as it still accepts a list as well
```suggestion
    def emit_logs(self, logs: list[AttributeDict] | tuple[AttributeDict] | AttributeDict, ignore: list = None) -> None:
```
Should add `from __future__ import annotations` to top of the file. I think we should start using this in favor of `typing.List` and the like as it is easier to read and cleaner.